### ISSUE TITLE
[UX] Move field's help text (description)

### ISSUE THREAD

**Issue by Graham-72** [backdrop/backdrop-issues#1403]
I would like to see a field's description/help text appear above the field rather than after, as would surely be the case with forms in everyday life (i.e. non-Drupal/Backdrop).

I have just read drupal.org/node/2319111 "Drupal 7 - Move field's help text (description) between the label and input element(s)". Do we have a Backdrop view or decision on this?

See https://github.com/backdrop/backdrop-issues/issues/1351 for one example.

**Comment by mikemccaffrey** [comment]
It would be hard to evaluate whether this is a good change to make, considering the thousands of different use cases of descriptive text with fields of different types. And since having the help text below the field is a long-standing drupal standard, we should probably not change it on a whim.

Perhaps we should consider alternate ways to resolve this issue, such as a new setting for whether the help text is is above or below the the field inputs, a contrib module that makes the change across the entire site for those who want it, or better visual indication that the help text is tied to the field above it.

**Comment by Graham-72** [comment]
Fair enough, though for me it is a bit more than a whim. It is a significant UX irritation.

**Comment by sutibun** [comment]
Generally speaking, **this is a UX fail.**

1. **Drupal/Bd is generally hard to understand and follow.** It has a long history of not being the most intuitive CMS. I doubt many can argue this point.
2. Knowing that, you know **users will eventually hit a UI element that will perplex them**. They won't understand what to make of it.
3. Now, because Drupal isn't well designed or user-friendly, you'll want, *at the very least*, to have the descriptive text **nearby** to explain what's going on. Having the help/descriptive text further away from the problem doesn't help this "painful" moment. You're now forcing the user to **look** around for help. This is what degrades the UX more than it should. A double whammy.

Here's a quick example from #1282 (think the search settings page?)

![search](https://cloud.githubusercontent.com/assets/14132746/10214112/a5185d5c-67da-11e5-815d-8334d76825f9.png)

What does "active search modules" mean? Do you think a user would easily grasp the meaning behind this title? Not likely. This will cause confusion. I know because it confused me when I first encountered this not too long ago. Yes, in Bd. (Theory: if it confuses even me, it's likely normal users won't stand a chance)

Let's start from the beginning again.

So, a user will first read a confusing title, then still not knowing what's what will then encounter two checkboxes and then finally hit the descriptive text that provides the context and understanding.

**Isn't that the wrong order?**

Shouldn't the user first understand what is going before encountering any UI element? Of course! Presenting UI elements before the user orients him/herself is a recipe for confusion.

This wouldn't be such a big trouble if for the most part, Drupal was easy to understand. But it's not. More often that not, users will find lots of quirky things that need an explanation. So, this is why it's best to move the help text near the label so you'll get the following process

1. User reads title/label but doesn't get it.
2. User eyes then falls to the help text which provides context and understanding.
3. User finally sees the UI element but because now have the proper background info, can follow along even if the UI is a lil' quirky.

I am in the fortunate (unfortunate?) position where I am not an advanced Drupal user so I am still tripping over Drupal/Bd and experiencing these pitfalls. I think if you know Drupal/Bd well, this doesn't seem like a big deal but I'd argue it is.

> And since having the help text below the field is **a long-standing drupal standard**, we should probably not change it on a whim.

"a long-standing drupal standard" That isn't a good argument. "Standard" makes it sound nice and authoritative... something not to challenge. But that is far from the truth. 

Standard just means it's what been done for ages BUT Drupal isn't exactly the most friendly CMS so more reason to change how it was done.

Having said that though, ultimately the best solution is to design Bd so it's more user friendlier and intuitive so we won't have to rely on help text. The moment you HAVE to use explain a design element, that's a sign you still have more work to do. To me help text is a crutch. But since it'll take time to refine Bd, moving the help text closer is a step in the right direction. A short term gap.

My two cents.

**Comment by Graham-72** [comment]
I share strong feelings about this. And it is not just about what one might call internal settings within Backdrop/Drupal. When building a site with quite complex data fields some help text is often needed to explain what the owner of the site must do to add content. Here is part of screenshot for a real life Backdrop site (a book store) where my client adds details about a book. Of course it is not a problem now that she has learnt where to look for the explanation, but it is not intuitive to look below the fields and requires supplementary notes for other staff in the book store.
![edit book the road to little dribbling - browsers bookshop 2015-12-15 04-52-51](https://cloud.githubusercontent.com/assets/10559274/11802521/8e4801be-a2e9-11e5-920f-76b67de27312.png)

**Comment by sutibun** [comment]
> Of course it is not a problem now that she has learnt where to look for the explanation, but it is not intuitive to look below the fields and requires supplementary notes for other staff in the book store.

What this means is she has adapted to Bd's quirkiness. Or really, just puts up with it. Although it might seem small but this tiny friction contributes to the feeling that Bd is just hard to work with. Little by little these things add up to overall frustration.

**Comment by klonos** [comment]
The underlying issue here is that different "level" of users expect different things...

- Newcomers would rather help text to be top-most so that they read an intro. Especially in pages that required deeper knowledge of Drupal/Backdrop internals, workflows and terminology.
- Veterans (or newcomers once they've learned a thing) would rather the text be out of their way, less page clutter and less scrolling.

So the question is: "Do we target newcomers or old-timers?" and as a side-question: "Can we find a way that sorts this issue for both?"

I do not have the answer for these questions. Perhaps as a veteran I get a bit selfish and tent to ignore the needs of people trying to learn Drupal/Backdrop in favor of getting things faster and having a cleaner UI.

Here's what this situation got me thinking though... in some video games there is a "Show tooltips" or "Enable tutorial" checkbox/mode. Can we have a "sticky" :question: help icon that gets the help text out of the way (completely - all of it!)?

**Comment by klonos** [comment]
...how about help icons next to each element that shows the help text as a tooltip on hover (click on mobile, touch devices)?

**Comment by klonos** [comment]
@Graham-72 having markdown links point to domain+path without the leading ```http://www``` (or simply the ```www``` alone) causes them to render like ```https://github.com/``` + GitHub path where the link was created + path entered, so your link was pointing to ```https://github.com/backdrop/backdrop-issues/issues/drupal.org/node/2319111``` which of course 404s. Fixed that for you :wink:

**Comment by klonos** [comment]
Just as a reference, there is https://www.drupal.org/project/beautytips that works by revealing a help text balloon/tooltip on field focus (personally don't like this solution because it does not give any visual indication that a help text is actually available). Also this one: https://www.drupal.org/project/helpinject that sorta does what I suggested above but with a "+" sign instead of a "?".

**Comment by klonos** [comment]
Here's a few more articles/screenshots of similar things to what I proposed:

https://ux.stackexchange.com/questions/53950/helptext-area-vs-help-tool-tip-icons-on-click

http://howto.caspio.com/datapages/datapage-components/field-configuration-options/rollover-hints/

https://www.drupal.org/sandbox/guillaumev/2074309

https://ux.stackexchange.com/questions/64313/how-to-show-static-user-initiated-and-dynamic-help-text-for-radio-buttons-and

https://civicrm.stackexchange.com/questions/3053/can-drupal-webform-integration-update-matching-contacts (the screenshot with the "?" next to the "existing contact" checkbox).

**Comment by Graham-72** [comment]
Thanks for all those references @klonos - plenty to think about.

Perhaps we should also look at what Wordpress offers its users. I don't know whether this is typical but how about this example taken from a narrow right-hand column:
![edit post smith imaging wordpress 2015-12-15 10-52-48](https://cloud.githubusercontent.com/assets/10559274/11808772/80a8be1e-a31a-11e5-92e8-2927c3813368.png)

**Comment by klonos** [comment]
You see, I understand that this provides the instructions for people that have not used this before, but once I have done it once and I am familiar with how to do it, then this rather lengthy text is in my way and simply adds clutter. I would like to get it out of my way. Now, if all this was to be "hidden away" under a :question: or :information_source: icon next to the field label and only shown once someone actually clicked there, it'd be perfect.

This design as is assumes that everyone would need this info available *all the time*. That is not true though. Think MS Office and the paperclip: most everyone though it has cute and helpful at first, but when you had it coming again and again telling you the same tips over and over while doing the same task, it got annoying.

**Comment by Graham-72** [comment]
Yes, perhaps it is a poor example of useful text (too much of it) but at least it is above the field.

**Comment by klonos** [comment]
> ...perhaps it is a poor example of useful text (too much of it)

Precisely why I'm not particularly font of inline help text. If we have a general standardized way of placement (either above or below elements) throughout the entire UI, although it might be ok with small portions of text, it gets really messy with big help text ...especially in small screens.

If you have strong feelings about getting the help text as close to the label of the fields, I believe that we can explore two different ways implementation-wise... either have the help text in a pop-up tooltip that pops close to the field or have it so that the help text is inline with the rest of the document (as opposed to having it in a separate pop-up dialogue) and so that clicking the help icon toggles the help text visibility (possibly with some animation a la fieldset or similar). Either way and anyway, I think that we should aim to simplify the UI by hiding help text and denoting its existence by icons that toggle visibility.

**Comment by klonos** [comment]
> If you have strong feelings about getting the help text as close to the label of the fields...

...never mind this particular comment of mine. After careful consideration, I think that it *is* best to have the help text between the label and the actual input elements in general (which is what this issue was filed for). Especially if one considers as an example use case a set of **many** checkboxes or radios. Having the help text after the input elements in such cases actually places it far from the eye of the user and constitutes a major UI fail.

I'll file a separate issue for my proposal to hide help text behind a help icon.

**Comment by klonos** [comment]
> And since having the help text below the field is a long-standing drupal standard, we should probably not change it on a whim.

...the overlay has been with us for quite a long time and can be considered a long-standing drupal standard UI-wise. I'm not sure you'll find a lot of people to say that it was a good thing though :wink:

**Comment by docwilmot** [comment]
> Perhaps we should also look at what Wordpress...

@klonos: https://github.com/backdrop/backdrop-issues/issues/1388 :grin: 

Also, our version sucks, whether its an established standard or not, it is poor UX. Was writing more but @klonos just intercepted my exact thoughts with that last post including the separate issue. So, ditto.

**Comment by klonos** [comment]
@docwilmot ...seems we were thinking the same thing at the same time.

**Comment by klonos** [comment]
...double ditto :smile:

**Comment by Graham-72** [comment]
Thanks guys. You have made me happy! 

> I'll file a separate issue for my proposal to hide help text behind a help icon.

I will add ot to my personal to-do list for 2016 - unless someone else gets there first.

**Comment by sutibun** [comment]
@klonos Nice. You widened the discussion with alternatives. Monkey was focusing too much on help text and only help text.

> Perhaps as a veteran I get a bit selfish and tent to ignore the needs of people trying to learn Drupal/Backdrop in favor of getting things faster and having a cleaner UI.

Monkey, the champion of the new user ^\_^

Anyway, the help text doesn't bother me even after I know how to use a page. But maybe others are different.

If it's a big issue, how about the following:

- By default, Bd should show the inline help text after the title/label. This is to help newcomers and less advanced users.
- Experienced users who no longer have a need for the help text can optionally go to a settings page and check off "Collapse all help text" which will remove the help text and replace with that question mark button.

**Comment by klonos** [comment]
I think that this implementation here would greatly improve the UX, but unfortunately not when on mobile. Please see the screenshot I've posted in https://github.com/backdrop/backdrop-issues/issues/1414#issuecomment-167260063 that shows that the clutter and need for long scrolling is there even after moving the help text between the label and the input. I strongly believe that after implementing this thing here, we should also implement #1414 and have it switch to that by default when we hit a mobile breakpoint.

**Comment by sutibun** [comment]
Didn't think about mobile. Good digging. Hmm... will have to agree then. So, how about we have this on the config page:

> How would you like to display the help text?
>
> - [x] **Default**: Show help text between the label and the input form. However, display a question mark button instead for small screens. (Recommended)
> - [ ] **Collapsed**: Instead of displaying help text, replace with a question mark button. Helps to reduce screen clutter.
> - [ ] **None**: Do not display any help text on the admin pages. Useful if you're no longer a beginner.

**Comment by jenlampton** [comment]
One thing I'd like to differentiate is when the help text is on a text field :
![screen shot 2016-01-06 at 11 12 58 pm](https://cloud.githubusercontent.com/assets/397895/12164540/36bf0160-b4cb-11e5-9db6-e8b37e54ca57.png)
vs when it's on radios or checkboxes:
![screen shot 2016-01-06 at 11 13 46 pm](https://cloud.githubusercontent.com/assets/397895/12164542/3a823c5e-b4cb-11e5-8608-9ec2661b9066.png)

I think for text fields it's fine where it is, since underneath the test field is not very "far" from the problem.  However when we are talking about checkboxes and radios, putting the description text underneath what could be a long list of options is certainly too "far" from the label.

Additionally, I've heard that putting the text way down there below radios/checkboxes is a uniquely Drupal problem (as in, no one else does that).  I've been staring at Drupal for so long now that I can't remember it being anywhere else. Do we have any other examples of radios or checkboxes from around the web?

I would be happy to introduce a change that cleans up our radios & checkboxes asap, and then we can continue to debate tooltips hover states and mobile interfaces :)

**Comment by Graham-72** [comment]
I think that is an excellent way forward. Thank you.
Here is an example from my online banking account which shows a mixture including tooltips.
![lloyds bank - make payment 2016-01-07 07-33-13](https://cloud.githubusercontent.com/assets/10559274/12164959/9be58712-b511-11e5-9453-33a2db60c35f.png)

**Comment by jenlampton** [comment]
I also like the description text above for textareas (which can also be large)
![screen shot 2016-01-06 at 11 42 20 pm](https://cloud.githubusercontent.com/assets/397895/12165054/4efff49c-b4cf-11e5-966b-41418ce55926.png)

**Comment by Graham-72** [comment]
> Long description

Yes please!

**Comment by docwilmot** [comment]
I am in support of moving it to just below the label for all fields, for consistency.

**Comment by klonos** [comment]
Reviving this old issue after 3 years 😄...

There is https://www.drupal.org/project/fel which offers a per-field option via the UI:

![element_order_fields_settings](https://user-images.githubusercontent.com/2423362/53357184-6768da00-3981-11e9-8cf0-5bb74483a500.png)

We could reduce this to a simple "Place help text after field label" checkbox, which could be less clutter IMO:

<img width="1193" alt="screen shot 2019-02-26 at 4 59 31 am" src="https://user-images.githubusercontent.com/2423362/53358091-79e41300-3983-11e9-9fc4-fe7ae3767923.png">

The good thing about that contrib module is that it also extends the Form API, by providing a `#description_display` property that we can use for our core admin forms, and contrib can also make use of it. So we can do:

```php
$form['example'] = array(
  '#type' => 'checkboxes',
  '#title' => t("So many checkboxes"),
  '#options' => $array_of_lots_and_lots_and_lots_of_options,
  '#description' => t("Normally, you'd have to scroll to see this, but it has now been moved after the field label ;)"),
  '#description_display' => 'before',
);
```

**Comment by klonos** [comment]
Turns out that this has been implemented in D8 core 👍 ...there is a new `#description_display` option with three allowed values: `before`, `after` and `invisible`. I have updated the issue title and summary, and I am making this part of #378

**Comment by herbdool** [comment]
I agree this should be in core. There is this d7 module https://www.drupal.org/project/label_help that is useful and easy to port (though it doesn't cover a number of field types). Perhaps providing a label help field is the way to go that would meet most people's needs.

**Comment by Graham-72** [comment]
I intended to do something about this in 2016 (see above). It is now 2020! I had forgotten all about it. But now I will assign it to myself and have it back on my personal agenda.

**Comment by Graham-72** [comment]
I have at long last produced an initial version of a PR - https://github.com/backdrop/backdrop/pull/3133 - that makes it possible for a site builder to specify for a specific field that the help text for that field (the description) appears before rather than after the field content, i.e. immediately after the field label.

**Comment by jenlampton** [comment]
Sanity-check Question: Is the position of the description text something that needs to be configured? I mean -- if core were to put it in the correct place to begin with. 

As it is now, the text is in the same place for every kind of element -- which we know isn't right. But! What if we fixed that? If we put the description text in a sensible location to start with, then I would think 'configurable positioning' should be left to contrib. I would prefer to fix a UX bug in core, than to add a new feature that adds a work-around for the bug. 

I thought there was an issue for this somewhere...  (maybe I'm remembering a D8 issue). I thought we had made a list of all the Form API element types, and decided where the description text should go for each one. But never gotten around to create theme function(s) to put it there? I'll see if I can find that.

**Comment by jenlampton** [comment]
I just re-read this whole thread to refresh my memory, and I see that I had the same opinion back in 2016 :)

> I also dislike the idea of asking a user where they want their help text. That's not a setting most people would even think about, so let's just try to get it right for them in core, without a UI :) Anyone who wants to change it can use a contrib module.

I like the Drupal 8 solution, and think we should do the same:

> Turns out that this has been implemented in D8 core 👍 ...there is a new `#description_display` option with three allowed values: `before`, `after` and `invisible`

And I am also (still) of the opinion that there should not be a UI for this in core. However! We should go through all the form API elements and set sensible defaults for 	`#description_display`, so that nobody ever needs to think about where the description text should go. 

@Graham-72 it looks like your PR adds a `#description_before` property instead of `#description_display`? Was that an intentional change from the D8 approach?

**Comment by Graham-72** [comment]
> @Graham-72 it looks like your PR adds a #description_before property instead of #description_display? Was that an intentional change from the D8 approach? (I'm asking because I'm not sure I see the value in having a 'invisible' option and wonder of a boolean TRUE/FALSE is preferred)

Yes, after working for some time with a `#description_display` property that included the 'invisible' option I realised that using a boolean property simplified the code and that there seems little point in the 'invisible' option. And so rather than use `#description_display` as a boolean I thought it better to change the name to `#description_before` to avoid confusion.

From what has been said previously, some have a strong preference for the help text always being after the field, which seems fine for regular users for whom it just provides visual clutter. But when providing a UI for occasional or first-time users I feel strongly that placing it before is more helpful and is what I would normally choose to do for my clients.

@jenlampton my personal view is that the option is sensible to include in core. It adds very little code and surely the additional UI complexity only exists when administering a content type, not to other site users.

**Comment by herbdool** [comment]
@Graham-72 your PR works fine, thanks for getting this started again. I didn't see any errors. Once there's consensus on the features I can do a more in-depth review of the code.

I was going to agree with @jenlampton's approach to put in some sensible defaults but don't expose it to the UI in core. But, then I thought, people might get confused about *when* the help text is appearing before or after. To them it might seem like it switches randomly. So I'm leaning to having it exposed in the UI, or putting the help text before for all fields by default.

The UI @Graham-72 added seems pretty minimal, much less intrusive than the CiviCRM field UI which has a "pre-help", "post-help" and "description".

**Comment by docwilmot** [comment]
> ...people might get confused about when the help text is appearing before or after. To them it might seem like it switches randomly. So I'm leaning to having it exposed in the UI...

Agreed with above; the UI option is unintrusive and minimal. And I agree completely that mysterious switching would prompt a user to try and figure out what controls the switching, and no UI would lead to more confusion.

**Comment by oadaeh** [comment]
> mysterious switching would prompt a user to try and figure out what controls the switching
And whether the description was for the field above it or below it (depending on other aspects of the UI, like spacing, field title display & position, etc.).

**Comment by oadaeh** [comment]
I'm curious @Graham-72, is

> But when providing a UI for occasional or first-time users I feel strongly that placing it before is more helpful and is what I would normally choose to do for my clients.

based on gut feeling, observation of users, feedback from clients, something else?

**Comment by jenlampton** [comment]
> @jenlampton my personal view is that the option is sensible to include in core. It adds very little code and surely the additional UI complexity only exists when administering a content type, not to other site users.

I disagree because I don't think this problem is limited to fields or content types. Form elements exist all over Backdrop and we need to fix the position of description text **everywhere**: on every single element, in every single form. 

I'd even argue that it's more important that we get the position of description text correct for administrators than for content entry, because *in most cases* there won't even be any description text on fields, but we certainly have it all over the place in core to try to explain how to use Backdrop to build a website :)

However, if most of you feel like we need it as part of Field UI, I could be convinced to have it there (only). Field UI is such a mess anyway I'm not sire one more option will make it much more daunting. :)

> To them it might seem like it switches randomly. 

I still disagree -- I expect that people won't notice where the text is is when they are able to use the form successfully. This is only noticeable now because we are currently doing it wrong almost everywhere :) 

I don't remember if the `correct` position of description text is something that is objective? Last time I looked into this (and it was some time ago...) I thought I found some UX guidelines about when it should go before and when it should go after a form element. Let me see if I can find that again...  (and if I can't -- maybe this memory is faulty..)

**Comment by jenlampton** [comment]
Well I just read several very interesting articles about form usability and learned some interesting things from eye-tracking results, but none of them specifically studied the position of description/help text on form elements. I'm going to check my google drive (like you can ever find anything in there...) for old Drupal usability study notes to see if maybe that's what I'm remembering.

**Comment by Graham-72** [comment]
@oadaeh 

> based on gut feeling, observation of users, feedback from clients, something else?

Perhaps mainly gut feeling, but from putting myself into the position of some of my users who have no experience of editing websites and are fearful of this new office task, and certainly no experience of Backdrop or Drupal, nor a need to learn any detail, just need easy-to-follow instructions on the screen, with 99% chance of getting it right without my intervention. These are not site builders, nor site administrators. They are content editors with very restricted permissions.

**Comment by klonos** [comment]
Here's a PR to get this going again: https://github.com/backdrop/backdrop/pull/4434

I didn't start off of the existing PR, because it was outdated, and since we want to support both `'before'`, `'after'` and `'invisible'` (same as in Drupal), whereas that older PR was adding `#description_before` explicitly.

The new PR has a line that intentionally flips all help text to be above the form elements (and for fieldsets where it is already above, it does the opposite). Please test and provide feedback.

**Comment by klonos** [comment]
> I am in support of moving it to just below the label for all fields, for consistency.

@docwilmot I know that this comment is from back in 2016, and that a lot might have changed since, but we are not 100% consistent as it is now. All fields/elements have their descriptions come at the end, except in fieldsets where the description comes right after the label (the legend). And I actually find that to be working better that way - here's a side-by-side before/after screenshot:
![image](https://user-images.githubusercontent.com/2423362/236667709-5faff2f4-561a-4816-a5b9-6f20465a4353.png)

Another thing to consider is radio and checkbox sets, as well as individual radios/checkboxes. I don't think that having the help text come before them works well visually (perhaps because of the lack of the bolded label text that other elements have 🤷🏼 ). For example, if you check the PR sandbox in `admin/config/search/settings`:
![image](https://user-images.githubusercontent.com/2423362/236664849-db66c4d8-cf5c-4251-bd88-770c622068e1.png)
...having said that, things may not work as great on narrow screens with lenghty help text, as it moves the element away from its label (but maybe that's a wider issue with mobile):
<img src="https://user-images.githubusercontent.com/2423362/236667154-871fbf2d-253e-4631-9e78-796e5c9b9da4.png" width=35%>

With radio/checkbox sets that have individual descriptions for each radio/checkbox, things certainly look odd if we place it `before`:
![image](https://user-images.githubusercontent.com/2423362/236667869-f71219b2-ad00-49e6-ab9b-7fd9c02c79c4.png)

**Comment by klonos** [comment]
...I'm thinking that perhaps these would be useful as theme settings - especially for admin themes, such as Seven. Perhaps I'll do that in the PR, to allow people to switch between `before`/`after`/`invisible` and test things.

**Comment by klonos** [comment]
Over in #6597 we have a valid use case in core where the help text that should apply to multiple radios would make better sense if it could be placed right after the label of the radio set:
![image](https://github.com/backdrop/backdrop-issues/assets/2423362/b0bda974-fafd-4dbc-bac3-7ba3e9c57733)

I'll try to loop back to this issue soon.

**Comment by herbdool** [comment]
Webform has had a "Description above field" option since, basically, forever.

![2024-06-17 09 48 09 bd3 lndo site 8b6c42a426fe](https://github.com/backdrop/backdrop-issues/assets/1071483/bd40acb4-c476-49b2-ad1f-2fb565e40d87)

**Comment by herbdool** [comment]
@klonos I was going to just suggest changes, but it ended up a bit bigger so here's a PR with suggestions. Feel free to update your PR.

If there are changes in the default `#description_display` they are put into `hook_element_info()` instead of adding a switch statement to `form.inc`. Makes it simpler and then also allows for other modules setting defaults. 

It now accounts for `text_format` as well. And it has basic tests.

I removed the theme overrides because I think it complicates things and I'm not sure about putting this in the theme settings. I think we should get the basics in first and discuss a UI separately.

**Comment by klonos** [comment]
Thanks @herbdool 🙏🏼 

> ... I'm not sure about putting this in the theme settings. I think we should get the basics in first and discuss a UI separately.

I merely did that as a means for people to be able to switch between the two options globally (in the PR sandbox) for testing/review purposes - never intended to ad it as part of this request here. I believe that the default should be left as it is currently (below/after), with a way for developers to switch to above/before on a case-by-case basis, via the Form API.

**Comment by herbdool** [comment]
@klonos [my PR](https://github.com/backdrop/backdrop-issues/issues/1403) now passes all tests.

**Comment by klonos** [comment]
I've closed my PR in favor of yours @herbdool ...thanks for picking this up 🙏🏼

**Comment by klonos** [comment]
@herbdool I've reviewed the PR code, and it looks good overall 👍🏼 🙏🏼 ...I've left a single comment with regards to allowing element titles to be empty (which I think we shouldn't be doing).

PS: Since your PR is based on mine, I don't think that I should be doing the final review or marking this RTBC. So, anyone else please?

**Comment by olafgrabienski** [comment]
> Note that a field set to Unlimited values will already show the description above the field so this PR doesn't change that.

Hm, as far as I see (with and without the PR), a field set to _Unlimited_ doesn't show the description above the field just because of the _Unlimited_ setting. Or maybe I misunderstood what you said, @herbdool.

**Comment by herbdool** [comment]
@olafgrabienski it depends on the field. Take a basic text field. If set to unlimited the help text is shown above like this:

![2024-07-18 14 35 48 bd3 lndo site c5b53aec3098](https://github.com/user-attachments/assets/f2297a04-d0a4-4be0-8af2-7a2b3f2dfa4e)

That's a screenshot from a site *without* this PR.

**Comment by stpaultim** [comment]
I'm interested in this issue. @herbdool do you have time to rebase the PR so that we can get a working sandbox?

**Comment by stpaultim** [comment]
I'm a bit surprised that I haven't chimed in on this issue earlier as it's been a big frustration when trying to build editor friendly user experiences. Help text often gets lost in the UI. 

I'm happy with the decision to simply provide better default positioning of help text, althought I could have also accepted a solution that providing configuration for this. 

I think this PR is going to be a big improvement. 

It seems that decision to put help text above or below the UI element is mostly based on proximity to the label. Larger elements have help text above them, while shorter elements have help text below. What about elements like image fields, where it's very cluttered below the element by default. 

In my screen shot, the custom help text gets lost amongst the default help text. Sure, it could make sense to lump all help text together. BUT, since the default help text regarding image sizing is always the same, I fear that the custom help text may be ignored and might also be better place above the field itself. 

![image](https://github.com/user-attachments/assets/778a3ede-3a9c-4ae2-bb1e-44743847c07f)

I don't want this specific decision to block this overall PR. If there is any debate about this idea, let's spin this off into it's own issue and merge the rest of this PR.

**Comment by stpaultim** [comment]
Just noticing that the current PR does not make position of #description configuration, as suggested by issue title. Current PR simply assigns better default positioning of help text. Should we change the title of this issue?

**Comment by herbdool** [comment]
@stpaultim I've edited the title. It's configurable via the API (with a new `#description_display` setting) but not in the UI.

**Comment by herbdool** [comment]
@stpaultim I think your comment about image field's default vs custom help text is a separate issue. It would require some other solution than what is proposed here, which deals with `#description` regardless of what the source is.

**Comment by stpaultim** [comment]
I've marked this WFM. 

Does someone want to be the advocate for this issue, so we can add it to the milestone.

https://docs.backdropcms.org/documentation/adding-milestones-to-issues

> **Minor release milestones**
> In order for an issue to remain tagged for the next minor release, it must meet all of the criteria below:
> 
> - It must have been tagged with the [milestone candidate - minor](https://github.com/backdrop/backdrop-issues/labels/milestone-candidate-minor) label.
> - It must have an advocate* (preferably, you!).
> - The advocate for this issue cannot also be the advocate for another open issue in this milestone, unless, of course, the other issue has a Pull Request in a near-ready state with either of the labels [pr - ready to be committed](https://github.com/backdrop/backdrop-issues/labels/pr-ready-to-be-committed) or [pr - works for me](https://github.com/backdrop/backdrop-issues/labels/pr-works-for-me).

**Comment by herbdool** [comment]
I was away from my desk for a few days. I'll advocate, but looks like it'll have to be for 1.30 at this point. @stpaultim

**Comment by quicksketch** [comment]
This looks really good @herbdool! I reviewed it and only found a few very minor things. Review in the PR: https://github.com/backdrop/backdrop/pull/4798#pullrequestreview-2425640831

Although fieldsets include descriptions, I had to confirm that [`#type = details` does not](https://docs.backdropcms.org/form_api#details). So no further changes needed there.

**Comment by herbdool** [comment]
@quicksketch ready for another review.

**Comment by quicksketch** [comment]
Looks good to me!

**Comment by quicksketch** [comment]
Wow, opened for nearly 10 years. Thank you everyone for adding to the discussion along the way. I've merged https://github.com/backdrop/backdrop/pull/4798 into 1.x for 1.30.0.

Lots of thanks to go around on this one. Of course @herbdool for bringing it to code and filing the PR, great work as always. Thank you @klonos, @stpaultim, and @olafgrabienski for testing (and earlier feedback). And thank you to our brainstormers @Graham-72, @jenlampton, @sutibun, @oadaeh, @klonos (again) and @docwilmot.
